### PR TITLE
Redis Client monitoring

### PR DESCRIPTION
**Pull request by mikkolaj**
Add possibility to monitor cluster state from the client's perspective.

### CODE REVIEW

File: redis/src/main/scala/com/avsystem/commons/redis/monitoring/SentinelStateObserver.scala [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.avsystem.commons
+package redis.monitoring
+
+import com.avsystem.commons.redis.RedisNodeClient
+
+trait SentinelStateObserver extends ConnectionStateObserver {
+  def onMasterChange(master: RedisNodeClient): Unit
```
Comment (mowczare) [inline redis/src/main/scala/com/avsystem/commons/redis/monitoring/SentinelStateObserver.scala:7]: Why do you need to return the whole client with its internals in this callback?

I'm missing the usage in https://gitlab.avsystem.com/csp/ump/ump/-/merge_requests/5096 so really hard to tell.

File: redis/src/main/scala/com/avsystem/commons/redis/monitoring/ClusterStateObserver.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.avsystem.commons
+package redis.monitoring
+
+/**
+  * Intended for monitoring client's connections.
+  * Should be non-blocking and handle internal exceptions by itself.
```
Comment (mowczare) [inline redis/src/main/scala/com/avsystem/commons/redis/monitoring/ClusterStateObserver.scala:6]: What about the other observers? I'm missing documentation of the public traits there.

File: redis/src/main/scala/com/avsystem/commons/redis/actor/SentinelsMonitoringActor.scala [hunk 3]
```diff
@@ -15,7 +16,8 @@ final class SentinelsMonitoringActor(
   seedSentinels: Seq[NodeAddress],
   config: MasterSlaveConfig,
   onInitFailure: Throwable => Unit,
-  onMasterChange: RedisNodeClient => Unit
+  onMasterChange: RedisNodeClient => Unit,
+  stateObserver: OptArg[SentinelStateObserver] = OptArg.Empty,
```
Comment (mowczare) [inline redis/src/main/scala/com/avsystem/commons/redis/actor/SentinelsMonitoringActor.scala:20]: That's a very specific `stateObserver`, I presume

**Review by mowczare — APPROVED** [review]

**Comment by mowczare** [comment]
LGTM after nits, and after the `SentinelStateObserver` API will be clear for me.

File: redis/src/main/scala/com/avsystem/commons/redis/monitoring/ConnectionStateObserver.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+package redis.monitoring
+
+import com.avsystem.commons.misc.{AbstractCase, AbstractNamedEnumCompanion, AutoNamedEnum}
+import com.avsystem.commons.redis.NodeAddress
+
+trait ConnectionStateObserver {
+  def onConnectionStateChange(addr: NodeAddress, state: ConnectionState): Unit
+}
+
+sealed abstract class ConnectionState extends AbstractCase with AutoNamedEnum
```
Comment (sebaciv) [inline redis/src/main/scala/com/avsystem/commons/redis/monitoring/ConnectionStateObserver.scala:11]: Use value enum

File: redis/src/main/scala/com/avsystem/commons/redis/actor/RedisConnectionActor.scala [hunk 5]
```diff
@@ -20,8 +21,11 @@ import scala.collection.mutable
 import scala.concurrent.duration.Duration
 import scala.util.Random
 
-final class RedisConnectionActor(address: NodeAddress, config: ConnectionConfig)
-  extends Actor with ActorLazyLogging { actor =>
+final class RedisConnectionActor(
+  address: NodeAddress,
+  config: ConnectionConfig,
+  connectionStateObserver: OptArg[ConnectionStateObserver] = OptArg.Empty
```
Comment (sebaciv) [inline redis/src/main/scala/com/avsystem/commons/redis/actor/RedisConnectionActor.scala:27]: Formatting, missing trailing comma

File: redis/src/main/scala/com/avsystem/commons/redis/monitoring/SentinelStateObserver.scala (hunk 1 above)
Comment (mikkolaj) [reply redis/src/main/scala/com/avsystem/commons/redis/monitoring/SentinelStateObserver.scala:7]: I'm 99% sure `NodeAddress` will suffice. Going to ask @pzareba22 though.

File: redis/src/main/scala/com/avsystem/commons/redis/monitoring/ClusterStateObserver.scala (hunk 2 above)
Comment (mikkolaj) [reply redis/src/main/scala/com/avsystem/commons/redis/monitoring/ClusterStateObserver.scala:6]: Added.

File: redis/src/main/scala/com/avsystem/commons/redis/actor/SentinelsMonitoringActor.scala (hunk 3 above)
Comment (mikkolaj) [reply redis/src/main/scala/com/avsystem/commons/redis/actor/SentinelsMonitoringActor.scala:20]: Changed to more specific `sentinelStateObserver`.

**Review by mowczare — APPROVED** [review]
LGTM

**Pull request merged**